contracts/TokenHolder.sol

```diff
@@ -441,18 +597,107 @@ contract TokenHolder is MultiSigWallet {
     }
 
     /**
-     * @dev Retrieves the first 4 bytes of input byte array into byte4.
-     *      Function requires:
-     *          - Input byte array's length is greater than or equal to 4.
+     * @notice Constructs data and performs verification of ephemeral key.
+     *
+     * @param _amount Redeem amount that will be transferred from tokenholder
+     *                account.
+     * @param _beneficiary The address in the origin chain where the tokens
+     *                     will be released.
+     * @param _gasPrice Gas price that tokenholder is ready to pay to get the
+     *                  redemption process done.
+     * @param _gasLimit Gas limit that tokenholder is ready to pay.
+     * @param _redeemerNonce Nonce of the redeemer address.
+     * @param _coGateway CoGateway contract address.
+     * @param _nonce The nonce of an ephemeral key that was used to sign
+     *               the transaction.
+     * @param _v V of the signature.
+     * @param _r R of the signature.
+     * @param _s S of the signature.
+     *
+     * @return ephemeralKey_ which is bool execution status of
+     *         coGateway.redeem.
      */
-    function bytesToBytes4(bytes _input) public pure returns (bytes4 out_) {
-        require(
-            _input.length >= 4,
-            "Input bytes length is less than 4."
+    function verifyRedeemExecutableTransaction(
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _redeemerNonce,
+        address _coGateway,
+        uint256 _nonce,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    )
+        private
+        returns (address ephemeralKey_)
+    {
+        bytes memory redeemData = abi.encodeWithSelector(
```

Comment:
I don't know if you're missing anything—it may be me who is missing something 😅 

What I understand is that `verifyExecutableTransaction` does not examine `redeemData` contents, but just concerns itself with whether the signature matches—literally anything could be redeemData. I see the sense in prescribing items needed to call CoGateway.redeem in redeemData; however, `REDEEM_SIGNED_DATA_CALLPREFIX` is not required for CoGateway.redeem. It seems like extra code/computation that adds no value; or, if I accept the argument that there is security value in having more components to redeemData, I question whether the value of this particular component is worth the confusion of implying that there is a function that matches this selector (which is not the case).

Please let me know if I've misunderstood the scope of `data` / how it is used in verification.

As far as `abi.encode`, that would be called with everything but REDEEM_SIGNED_DATA_CALLPREFIX, so maybe that's why you are getting an error?

Reply:
Your understanding of redeemData is correct.

To generate messageHash as per EIP1077 proposal, dataHash is needed. 
REDEEM_SIGNED_DATA_CALLPREFIX is being used here to construct dataHash.

I understand your view point of should we have REDEEM_SIGNED_DATA_CALLPREFIX or not. 

Below are more convincing points which adds value to this approach : 

- Facilitator calls TH.redeem. Signing the data with _amount, _beneficiary, _gasPrice, _gasLimit, _redeemerNonce makes sure bad facilitator can't modify these arguments.

- This approach is also consistent with stake where facilitator calls stake and messageHash is generated with incoming arguments.

- I am not sure about exact gas cost for abi.encodeWithSelector. However I don't think it will not be too large considering keccak256 takes 1000 wei. If you really think it's a concern I can get the gas cost for this function. 

- Thinking around this from user point of view, user knows about _amount, _beneficiary, _gasPrice, _gasLimit, _redeemerNonce and can happily sign this data 😄  

Also I am not sure on any alternative approach. Please suggest if possible with code examples 👨‍💻